subprojects/build-operations/src/main/java/org/gradle/trace/pid/PidCollectorBuildService.java

```diff
@@ -0,0 +1,53 @@
+package org.gradle.trace.pid;
+
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.internal.GradleInternal;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.gradle.build.event.BuildEventsListenerRegistry;
+import org.gradle.tooling.events.FinishEvent;
+import org.gradle.tooling.events.OperationCompletionListener;
+import org.gradle.util.GFileUtils;
+
+import java.io.File;
+import java.lang.management.ManagementFactory;
+import java.lang.management.RuntimeMXBean;
+import java.lang.reflect.InvocationTargetException;
+
+public abstract class PidCollectorBuildService implements BuildService<PidCollectorBuildService.Parameters>, OperationCompletionListener {
+
+    public static void registerBuildService(GradleInternal gradle, File outFile) {
+        Provider<PidCollectorBuildService> pidCollectorService = gradle.getSharedServices().registerIfAbsent("pidCollector", PidCollectorBuildService.class, spec -> spec.parameters(
+            params -> params.getPidFile().set(outFile)
+        ));
+        gradle.getServices().get(BuildEventsListenerRegistry.class).onTaskCompletion(pidCollectorService);
+    }
+
+    public interface Parameters extends BuildServiceParameters {
+        RegularFileProperty getPidFile();
+    }
+
+    public PidCollectorBuildService() throws Exception {
+        GFileUtils.writeFile(getPid().toString(), getParameters().getPidFile().get().getAsFile());
+    }
+
+    private Integer getPid() throws NoSuchFieldException, IllegalAccessException, NoSuchMethodException, InvocationTargetException {
```

Comment:
I am not very happy with how we obtain the PID of the current process here, though I don't have a better idea :(.

Reply:
I guess no better solution on Java 8.